tests/math/test_matrix_manipulation.py

```diff
@@ -1068,3 +1070,195 @@ def test_expand_vector_invalid_vector(self):
+            ValueError,
+            "Factor is exactly singular",
+        )
+    ]
+
+    @pytest.mark.parametrize("dm", dm_list)
+    def test_sqrt_matrix_sparse_dm(self, dm, tol):
+        """Test the sqrt_matrix function."""
+
+        A = qml.math.asarray(dm)
+        A_sparse = csr_matrix(A)
+
+        assert np.allclose(sqrt_matrix(A), sqrt_matrix_sparse(A_sparse).toarray(), atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("shape", shape_list)
+    def test_sqrt_matrix_sparse_random(self, shape, tol):
+        """Test the sqrt_matrix function."""
+
+        # From unitary group
+        dm = unitary_group.rvs(shape)
+        dm = dm @ dm.T.conj()
+        A = qml.math.asarray(dm)
+        A_sparse = csr_matrix(A)
+
+        assert np.allclose(sqrt_matrix(A), sqrt_matrix_sparse(A_sparse).toarray(), atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("matrix", matrices)
+    def test_sqrt_matrix_inputs(self, matrix, tol):
+        """Test sqrt_matrix function with various input matrices."""
+
+        # type M^† M
+        A = matrix.T.conj() @ matrix
+        A = np.eye(A.shape[0]) - A
+        A_sparse = csr_matrix(A)
+
+        result_sparse = sqrt_matrix_sparse(A_sparse)
+        result = result_sparse.toarray()
+        result_2 = result @ result
+
+        assert np.allclose(result_2, A, atol=tol, rtol=0)
+
+        # type M M^†
+        A = matrix @ matrix.T.conj()
+        A = np.eye(A.shape[0]) - A
+        A_sparse = csr_matrix(A)
+
+        result_sparse = sqrt_matrix_sparse(A_sparse)
+        result = result_sparse.toarray()
+        result_2 = result @ result
+
+        assert np.allclose(result_2, A, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("m", matrices_negative)
+    def test_sqrt_matrix_sparse_input_negative(self, m):
+        """Test that sqrt_matrix_sparse raises ValueError for matrices with negative eigenvalues."""
+        m = csr_matrix(m)
+        with pytest.raises(ValueError):
+            sqrt_matrix_sparse(m)
+
+    @pytest.mark.parametrize("m, e, info", illmats_info_pairs)
+    def test_sqrt_matrix_sparse_input_ill_conditioned(self, m, e, info):
+        """Test that appropriate errors are raised for ill-conditioned matrices"""
+        m = csr_matrix(m)
+        with pytest.raises(e, match=info):
+            sqrt_matrix_sparse(m)
+
+    def test_sqrt_matrix_sparse_input_valid(self):
+        """Test that if dense input errors raised"""
+        A = np.array([[1, 0], [0, 1]])
+        with pytest.raises(TypeError, match="only supports sparse matrices"):
+            sqrt_matrix_sparse(A)
+
+
+class TestDenmanBeaversIterations:
+    """Tests for the Denman-Beavers iteration method for matrix square root"""
+
+    def test_singular_matrix(self):
+        """Test that singular matrix raises appropriate error"""
+        n = 4
+        mat = csr_matrix(np.diag([0.0] + [1.0] * (n - 1)))
+
+        with pytest.raises(ValueError, match="Factor is exactly singular"):
+            _denman_beavers_iterations(mat)
```

Comment:
While fine for now, I think if we can abstract away details of private functions behind an API we should.
For this, I think defining a `solver` or something similar interface that allows us to choose this implementation (or a more stable/convergent one) could be a better strategy in the long run.

Maybe taking note for future modifications in the backlog would be worthwhile.

Reply:
I'm also thinking about that! For this sqrtm functionalities there are quite some other ways to implement, even within the same scope of iterative methods. I'll make backlog and tag back

Reply:
sc-85710